# Uncovering Furniture Sales Insights

We, Sheetal Jain and Shivani Sharma conduct detailed Exploratory Data Analysis (EDA) on furniture sales dataset expanding over 2017-2020. Our dataset encompasses key dimensions such as Sales Order Number, Order Date, Delivery Date, Ship Mode, Customer Key, Product Key, and more. Through meticulous examination, we aim to unveil trends, patterns, and invaluable insights that shed light on the dynamics of furniture sales.

# Introduction

we delve into the depths of Exploratory Data Analysis (EDA), armed with a rich dataset encompassing vital dimensions.

With a robust dataset teeming with essential dimensions like Sales Order Number, Order Date, Delivery Date, Ship Mode, Customer Key, Product Key, Category Name, Subcategory Name, and more, we embark on common EDA tasks of data cleaning, feature engineering, visualization, and statistical analysis. Through meticulous scrutiny and insightful analysis, we are aimed at gaining comprehensive insights that empower strategic decision-making in the dynamic world of furniture retail.

# Importing all the necessary Python modules 

In [1]:
#Numpy: For Numerical Computation
import numpy as np
#Pandas: For Data Cleaning,Selection,Filtering,Aggregation,transformation,visualization,Analysis
import pandas as pd
#Math :  for using mathematical formulas 
import math
#Matplotlib: for representing data in graphical form
import matplotlib as plt

# Importing Data Set (Data Retrieval)

Since dataset is in current directry, we are passing file name as it is.

In [2]:
furn_df=pd.read_excel('Office_Sales.xlsx')

### Converting Excel files to CSV (Comma Separated Values) format 

Working with CSV file has several benefits such as simplicity, portability, compatiability, reduced file size, easier integration, data integration, etc. So, we will convert our file in csv before working with it.

In [3]:
furn_df.to_csv('Furniture_sales_data_set.csv')

In [4]:
# fdf stands for furniture data frame
fdf = pd.DataFrame(furn_df)

### Verfiying whether the data frame has been created or not

In [5]:
fdf

SalesOrderNumber  OrderDate DeliveryDate        ShipMode  CustomerKey  \
0        IN-2017-47883 2017-01-01   2017-01-08  Standard Class           56   
1         IZ-2017-4680 2017-01-03   2017-01-07  Standard Class           66   
2        ID-2017-80230 2017-01-03   2017-01-09  Standard Class           64   
3        ID-2017-80230 2017-01-03   2017-01-09  Standard Class           64   
4      ES-2017-4869686 2017-01-03   2017-01-07  Standard Class           63   
...                ...        ...          ...             ...          ...   
29994    ID-2020-63864 2020-12-30          NaT  Standard Class        18820   
29995     PL-2020-1210 2020-12-30          NaT  Standard Class        18833   
29996   MX-2020-110527 2020-12-31          NaT    Second Class        18654   
29997  ES-2020-2815584 2020-12-31          NaT  Standard Class        11154   
29998    IN-2020-75603 2020-12-31          NaT    Second Class        18846   

             ProductKey     CategoryName SubcategoryName  \
0       FUR-FU-10003447        Furniture     Furnishings   
1      FUR-NOV-10002791        Furniture          Chairs   
2       FUR-CH-10000666        Furniture          Chairs   
3       FUR-CH-10000214        Furniture          Chairs   
4       FUR-BO-10000728        Furniture       Bookcases   
...                 ...              ...             ...   
29994   OFF-SU-10002388  Office Supplies        Supplies   
29995  OFF-KLE-10001644  Office Supplies        Supplies   
29996   OFF-SU-10001624  Office Supplies        Supplies   
29997   OFF-SU-10004279  Office Supplies        Supplies   
29998   OFF-SU-10001618  Office Supplies        Supplies   

                              ProductName  UnitPrice  OrderQuantity  \
0              Eldon Light Bulb, Duo Pack      25.26              5   
1        Novimex Swivel Stool, Set of Two     166.71              4   
2                   SAFCO Chairmat, Black      57.39              2   
3           Hon Rocking Chair, Set of Two     132.87              2   
4      Dania Corner Shelving, Traditional     122.07              7   
...                                   ...        ...            ...   
29994             Elite Shears, Easy Grip      48.63              2   
29995             Kleencut Trimmer, Steel      40.83              6   
29996          Kleencut Ruler, High Speed       9.04              8   
29997           Kleencut Shears, Serrated      43.47              6   
29998   Kleencut Letter Opener, Easy Grip      28.68              9   

       Discount %  ShippingCost OrderPriority  
0             0.5         11.92        Medium  
1             0.6          9.81          High  
2             0.2          8.30           Low  
3             0.0          9.63           Low  
4             0.6         12.56        Medium  
...           ...           ...           ...  
29994         0.7          3.25        Medium  
29995         0.0         34.18        Medium  
29996         0.5         10.75        Medium  
29997         0.7          3.67          High  
29998         0.8         14.20        Medium  

[29999 rows x 14 columns]

### Description of dataset

##### The furniture sales dataset contains the following information for each of the 29,999 sales orders:

`Sales order number:` This unique identifier ranges from IN-2017-47883 to PL-2020-1210.<br>
`Order date:` This indicates the date the order was placed, ranging from 2017-01-03 to 2020-12-30.<br>
`Delivery date:` This specifies the date the order was delivered, with some entries missing delivery dates (NaT).<br>
`Ship mode:` This refers to the shipping method used, with "Standard" being the most frequent option.<br>
`Customer key:` This seems to be a unique identifier for the customer who placed the order, ranging from 56 to 18859.<br>
`Product key:` This likely identifies the specific product sold, ranging from FUR-FU-10003447 to OFF-KLE-10001644.<br>
`Category name:` This broad category groups the product, such as "Furniture" or "Office Supplies".<br>
`Subcategory name:` This provides more specific details about the product category, such as "Furnishings" or "Chairs".<br>
`Product name:` This is the name of the product sold, like "Eldon Light Bulb, Duo Pack" or "Novimex Swivel Stool. Set of Two".<br>
`Unit price:` This shows the price per unit of the product, ranging from $1.39 to $1889.99.<br>
`Order quantity:` This specifies the number of units purchased in each order, ranging from 1 to 16.<br>

# Overview of furniture data set

## Column name and the number of unique elements in that column

In [6]:
pd.Series(fdf.nunique())

SalesOrderNumber    17801
OrderDate            1406
DeliveryDate         1455
ShipMode                4
CustomerKey         16874
ProductKey           5963
CategoryName            2
SubcategoryName        10
ProductName          2432
UnitPrice            2796
OrderQuantity          16
Discount %              9
ShippingCost        11144
OrderPriority           4
dtype: int64

## Overview of numerical column

In [7]:
fdf.describe()

CustomerKey     UnitPrice  OrderQuantity    Discount %  ShippingCost
count  29999.000000  29999.000000   29999.000000  29999.000000  29999.000000
mean    9154.332778     79.367678       3.476916      0.393013     30.353239
std     5412.576391    121.566872       2.284369      0.259108     68.112116
min       56.000000      1.390000       1.000000      0.000000      0.002000
25%     4424.000000     15.750000       2.000000      0.200000      3.630000
50%     9043.000000     33.560000       3.000000      0.400000      9.509000
75%    13784.000000     82.230000       5.000000      0.600000     27.211500
max    18859.000000   1889.990000      16.000000      0.800000   1581.510000

##  The Available Shipping Modes, Furniture Categories, Furniture Sub Categories, Products, Discount Rates, and Priority types in the data set

### `Shipping Modes`

In [8]:
shippingmode=pd.Series(fdf['ShipMode'].unique())
shippingmode

0    Standard Class
1      Second Class
2       First Class
3          Same Day
dtype: object

### `Furniture Categories`

In [9]:
furn_categories=pd.Series(fdf['CategoryName'].unique())
furn_categories

0          Furniture
1    Office Supplies
dtype: object

### `Furniture Sub Categories`

In [10]:
furn_sub_categories=pd.Series(fdf['SubcategoryName'].unique())
furn_sub_categories

0    Furnishings
1         Chairs
2      Bookcases
3         Tables
4     Appliances
5            Art
6        Binders
7          Paper
8        Storage
9       Supplies
dtype: object

### `Products`

In [11]:
products=pd.Series(fdf['ProductName'].unique())
products

0                              Eldon Light Bulb, Duo Pack
1                        Novimex Swivel Stool, Set of Two
2                                   SAFCO Chairmat, Black
3                           Hon Rocking Chair, Set of Two
4                      Dania Corner Shelving, Traditional
                              ...                        
2427                          Kleencut Shears, High Speed
2428              Acme Preferred Stainless Steel Scissors
2429                                       Letter Slitter
2430    Kleencut Forged Office Shears by Acme United C...
2431                  Acme Kleencut Forged Steel Scissors
Length: 2432, dtype: object

#### There are total  2432 products in furniture data set

### `Discount Rates`

In [12]:
dsct_rates=pd.Series(fdf['Discount %'].unique())
dsct_rates

0    0.5
1    0.6
2    0.2
3    0.0
4    0.7
5    0.8
6    0.3
7    0.4
8    0.1
dtype: float64

### `Priority Types`

In [13]:
order_priority=pd.Series(fdf['OrderPriority'].unique())
order_priority

0      Medium
1        High
2         Low
3    Critical
dtype: object

# Data Cleansing 

## Removing Missing Values

In [14]:
missing_values_count = fdf.isna().sum()
missing_values_count

SalesOrderNumber      0
OrderDate             0
DeliveryDate        170
ShipMode              0
CustomerKey           0
ProductKey            0
CategoryName          0
SubcategoryName       0
ProductName           0
UnitPrice             0
OrderQuantity         0
Discount %            0
ShippingCost          0
OrderPriority         0
dtype: int64

### We can see that there are 170 missing values in DeliveryDate column. To handle missing dates we have following options:
* Use a Default Date
* Use the Most Frequent Date
* Use Forward or Backward Fill
* Interpolate Dates
* Drop Missing Dates
* Use Domain Knowledge

##### For tracking undelivered records we can separated them and store in a different file.

### Separating records of Undelivered Orders and Saving them in a different file

In [15]:
#Separating 
Undelivered_orders = fdf[fdf.isnull().any(axis=1)]
Undelivered_orders.to_excel('undelivered_orders.xlsx')
#View of these records
Undelivered_orders

SalesOrderNumber  OrderDate DeliveryDate        ShipMode  CustomerKey  \
8489    CA-2020-128734 2020-12-25          NaT  Standard Class        18679   
8490    CA-2020-128734 2020-12-25          NaT  Standard Class        18679   
8491      AG-2020-2280 2020-12-25          NaT  Standard Class        18705   
8494      AG-2020-2280 2020-12-25          NaT  Standard Class        18705   
8506   IT-2020-4907444 2020-12-26          NaT  Standard Class        18731   
...                ...        ...          ...             ...          ...   
29994    ID-2020-63864 2020-12-30          NaT  Standard Class        18820   
29995     PL-2020-1210 2020-12-30          NaT  Standard Class        18833   
29996   MX-2020-110527 2020-12-31          NaT    Second Class        18654   
29997  ES-2020-2815584 2020-12-31          NaT  Standard Class        11154   
29998    IN-2020-75603 2020-12-31          NaT    Second Class        18846   

             ProductKey     CategoryName SubcategoryName  \
8489    FUR-CH-10001394        Furniture          Chairs   
8490    FUR-FU-10001731        Furniture     Furnishings   
8491   FUR-SAU-10000893        Furniture       Bookcases   
8494   FUR-SAF-10002121        Furniture          Chairs   
8506    FUR-CH-10000214        Furniture          Chairs   
...                 ...              ...             ...   
29994   OFF-SU-10002388  Office Supplies        Supplies   
29995  OFF-KLE-10001644  Office Supplies        Supplies   
29996   OFF-SU-10001624  Office Supplies        Supplies   
29997   OFF-SU-10004279  Office Supplies        Supplies   
29998   OFF-SU-10001618  Office Supplies        Supplies   

                               ProductName  UnitPrice  OrderQuantity  \
8489        Global Leather Executive Chair     350.99              3   
8490     Acrylic Self-Standing Desk Frames       2.67              4   
8491     Sauder Stackable Bookrack, Mobile     146.82              2   
8494            SAFCO Rocking Chair, Black     132.48              2   
8506   Harbour Creations Bag Chairs, Black      59.19              1   
...                                    ...        ...            ...   
29994              Elite Shears, Easy Grip      48.63              2   
29995              Kleencut Trimmer, Steel      40.83              6   
29996           Kleencut Ruler, High Speed       9.04              8   
29997            Kleencut Shears, Serrated      43.47              6   
29998    Kleencut Letter Opener, Easy Grip      28.68              9   

       Discount %  ShippingCost OrderPriority  
8489          0.8         29.83        Medium  
8490          0.8         15.57        Medium  
8491          0.7         26.91        Medium  
8494          0.2          1.03        Medium  
8506          0.8          4.01           Low  
...           ...           ...           ...  
29994         0.7          3.25        Medium  
29995         0.0         34.18        Medium  
29996         0.5         10.75        Medium  
29997         0.7          3.67          High  
29998         0.8         14.20        Medium  

[170 rows x 14 columns]

### Setting DeliveryDate cell 'In-Process' for Undelivered Orders or Missing Values Cells

In [16]:
fdf['DeliveryDate'] = fdf['DeliveryDate'].fillna('In-Process')        

In [17]:
fdf

SalesOrderNumber  OrderDate         DeliveryDate        ShipMode  \
0        IN-2017-47883 2017-01-01  2017-01-08 00:00:00  Standard Class   
1         IZ-2017-4680 2017-01-03  2017-01-07 00:00:00  Standard Class   
2        ID-2017-80230 2017-01-03  2017-01-09 00:00:00  Standard Class   
3        ID-2017-80230 2017-01-03  2017-01-09 00:00:00  Standard Class   
4      ES-2017-4869686 2017-01-03  2017-01-07 00:00:00  Standard Class   
...                ...        ...                  ...             ...   
29994    ID-2020-63864 2020-12-30           In-Process  Standard Class   
29995     PL-2020-1210 2020-12-30           In-Process  Standard Class   
29996   MX-2020-110527 2020-12-31           In-Process    Second Class   
29997  ES-2020-2815584 2020-12-31           In-Process  Standard Class   
29998    IN-2020-75603 2020-12-31           In-Process    Second Class   

       CustomerKey        ProductKey     CategoryName SubcategoryName  \
0               56   FUR-FU-10003447        Furniture     Furnishings   
1               66  FUR-NOV-10002791        Furniture          Chairs   
2               64   FUR-CH-10000666        Furniture          Chairs   
3               64   FUR-CH-10000214        Furniture          Chairs   
4               63   FUR-BO-10000728        Furniture       Bookcases   
...            ...               ...              ...             ...   
29994        18820   OFF-SU-10002388  Office Supplies        Supplies   
29995        18833  OFF-KLE-10001644  Office Supplies        Supplies   
29996        18654   OFF-SU-10001624  Office Supplies        Supplies   
29997        11154   OFF-SU-10004279  Office Supplies        Supplies   
29998        18846   OFF-SU-10001618  Office Supplies        Supplies   

                              ProductName  UnitPrice  OrderQuantity  \
0              Eldon Light Bulb, Duo Pack      25.26              5   
1        Novimex Swivel Stool, Set of Two     166.71              4   
2                   SAFCO Chairmat, Black      57.39              2   
3           Hon Rocking Chair, Set of Two     132.87              2   
4      Dania Corner Shelving, Traditional     122.07              7   
...                                   ...        ...            ...   
29994             Elite Shears, Easy Grip      48.63              2   
29995             Kleencut Trimmer, Steel      40.83              6   
29996          Kleencut Ruler, High Speed       9.04              8   
29997           Kleencut Shears, Serrated      43.47              6   
29998   Kleencut Letter Opener, Easy Grip      28.68              9   

       Discount %  ShippingCost OrderPriority  
0             0.5         11.92        Medium  
1             0.6          9.81          High  
2             0.2          8.30           Low  
3             0.0          9.63           Low  
4             0.6         12.56        Medium  
...           ...           ...           ...  
29994         0.7          3.25        Medium  
29995         0.0         34.18        Medium  
29996         0.5         10.75        Medium  
29997         0.7          3.67          High  
29998         0.8         14.20        Medium  

[29999 rows x 14 columns]

### Verifying if there are anymore missing values

In [18]:
missing_values_count = fdf.isna().sum()
missing_values_count

SalesOrderNumber    0
OrderDate           0
DeliveryDate        0
ShipMode            0
CustomerKey         0
ProductKey          0
CategoryName        0
SubcategoryName     0
ProductName         0
UnitPrice           0
OrderQuantity       0
Discount %          0
ShippingCost        0
OrderPriority       0
dtype: int64

`It's evident that we no longer have any missing values. Moving forward, our next project phase entails visualizing the furniture dataset.`

# We are aiming to perform analysis for following:
* Distribution of sales orders over time
* Different shipping modes comparison in terms of frequency and total sales
* Average order quantity for each product category and subcategory
* The distribution of unit prices across different product categories
* Average shipping cost per order for each shipping mode
* Correlations between order quantity and unit price
* The discount rate affect on the total sales amount
* The distribution of order priorities among the sales orders
* Varying sales across different days of the week or months or year
* Outliers in the shipping cost or unit price data
* The average delivery time for orders placed under each shipping mode
* Is there any correlation between the customer key and the total sales amount
* Sales difference between furniture and office supplies categories
* Patterns in the order quantities of specific products over time for different furniture categories
* Products that consistently have higher or lower discounts
* Varying order priorities across different product categories
* The relationship between order date and delivery date and pattern of delivery time for different products
* Varying order quantities for different order priorities